packages/@react-aria/overlays/src/calculatePosition.ts

```diff
@@ -457,3 +468,54 @@ function getPosition(node: Element, parent: Element): Offset {
   offset.left -= parseInt(style.marginLeft, 10) || 0;
   return offset;
 }
+
+// Returns the containing block of an element, which is the element that
+// this element will be positioned relative to.
+// https://developer.mozilla.org/en-US/docs/Web/CSS/Containing_block
+function getContainingBlock(node: HTMLElement): Element {
+  // The offsetParent of an element in most cases equals the containing block.
+  // https://w3c.github.io/csswg-drafts/cssom-view/#dom-htmlelement-offsetparent
+  let offsetParent = node.offsetParent;
+
+  // The offsetParent algorithm terminates at the document body,
+  // even if the body is not a containing block. Double check that
+  // and use the documentElement if so.
+  if (
+    offsetParent &&
+    offsetParent === document.body &&
+    window.getComputedStyle(offsetParent).position === 'static' &&
+    !isContainingBlock(offsetParent)
+  ) {
+    offsetParent = document.documentElement;
+  }
+
+  // TODO(later): handle table elements?
+
+  // The offsetParent can be null if the element has position: fixed, or a few other cases.
+  // We have to walk up the tree manually in this case because fixed positioned elements
+  // are still positioned relative to their containing block, which is not always the viewport.
+  if (offsetParent == null) {
+    offsetParent = node.parentElement;
+    while (offsetParent && !isContainingBlock(offsetParent)) {
+      offsetParent = offsetParent.parentElement;
+    }
+  }
+
+  // Fall back to the viewport.
+  return offsetParent || document.documentElement;
+}
+
+// https://developer.mozilla.org/en-US/docs/Web/CSS/Containing_block#identifying_the_containing_block
+function isContainingBlock(node: Element): boolean {
+  let style = window.getComputedStyle(node);
```

Comment:
It wouldn't be every render, just when we are repositioning. And in most cases there will be a single call - the times when `offsetParent` is null are rare.

I also think it will be exceedingly rare that the container of an overlay is a table so I left that for later.